give a better error when rustdoc exits with a non-zero exit code

### Problem

currently, the error message when rustdoc errors looks like this:
```
; RUSTDOCFLAGS=-Dwarnings cargo doc --no-deps --document-private-items --message-format short -q
src/main.rs:1:11: error: unclosed HTML tag `usize`
error: could not document `example`

Caused by:
  process didn't exit successfully: `/Users/jyn/.local/lib/rustup/toolchains/nightly-aarch64-apple-darwin/bin/rustdoc --edition=2021 --crate-type bin --crate-name example src/main.rs -o /Users/jyn/.local/lib/cargo/target/doc --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat,diagnostic-short --diagnostic-width=186 --document-private-items '-Arustdoc::private-intra-doc-links' -C metadata=a59e1929dfc569b0 -L dependency=/Users/jyn/.local/lib/cargo/target/debug/deps --extern bitflags=/Users/jyn/.local/lib/cargo/target/debug/deps/libbitflags-d34f162dfbd7ed32.rmeta --extern inner=/Users/jyn/.local/lib/cargo/target/debug/deps/libinner-a5a332b9f54b66cc.rmeta -Dwarnings --crate-version 0.1.0` (exit status: 1)
```

this is already kind of long, and it gets much worse as you add more dependencies.

### Steps

1. add some docs that will warn, e.g. `//! Option<usize>` at the root of the crate
2. `RUSTDOCFLAGS=-Dwarnings cargo doc`

### Possible Solution(s)

it would be nice to make it consistent with the error for rustc:
```
; cargo c -q --message-format short
src/main.rs:10:12: error[E0038]: the trait `Bar` cannot be made into an object
src/main.rs:10:32: error[E0038]: the trait `Bar` cannot be made into an object
error: could not compile `example` (bin "example") due to 2 previous errors
```

### Notes

_No response_

### Version

```text
cargo 1.77.0-nightly (ac6bbb332 2023-12-26)
release: 1.77.0-nightly
commit-hash: ac6bbb33293d8d424c17ecdb42af3aac25fb7295
commit-date: 2023-12-26
host: aarch64-apple-darwin
libgit2: 1.7.1 (sys:0.18.1 vendored)
libcurl: 8.4.0 (sys:0.4.70+curl-8.5.0 system ssl:(SecureTransport) LibreSSL/3.3.6)
ssl: OpenSSL 1.1.1w  11 Sep 2023
os: Mac OS 14.3.0 [64-bit]
```

the message format thing is because i used a bad example; rustdoc's error is a denied lint, which doesn't have an error code. i can change it to be more consistent with slightly different code:
```
; head -n3 src/main.rs
const foo: size = 1;
; cargo doc --no-deps --message-format short
src/main.rs:3:12: error[E0412]: cannot find type `size` in this scope
```

--- a/src/cargo/core/compiler/mod.rs
+++ b/src/cargo/core/compiler/mod.rs
@@ -376,19 +376,6 @@ fn rustc(cx: &mut Context<'_, '_>, unit: &Unit, exec: &Arc<dyn Executor>) -> Car
             }
         }
 
-        fn verbose_if_simple_exit_code(err: Error) -> Error {
-            // If a signal on unix (`code == None`) or an abnormal termination
-            // on Windows (codes like `0xC0000409`), don't hide the error details.
-            match err
-                .downcast_ref::<ProcessError>()
-                .as_ref()
-                .and_then(|perr| perr.code)
-            {
-                Some(n) if cargo_util::is_simple_exit_code(n) => VerboseError::new(err).into(),
-                _ => err,
-            }
-        }
-
         state.running(&rustc);
         let timestamp = paths::set_invocation_time(&fingerprint_dir)?;
         if build_plan {
@@ -507,6 +494,19 @@ fn rustc(cx: &mut Context<'_, '_>, unit: &Unit, exec: &Arc<dyn Executor>) -> Car
             }
         }
         Ok(())
+    }
+}
+
+fn verbose_if_simple_exit_code(err: Error) -> Error {
+    // If a signal on unix (`code == None`) or an abnormal termination
+    // on Windows (codes like `0xC0000409`), don't hide the error details.
+    match err
+        .downcast_ref::<ProcessError>()
+        .as_ref()
+        .and_then(|perr| perr.code)
+    {
+        Some(n) if cargo_util::is_simple_exit_code(n) => VerboseError::new(err).into(),
+        _ => err,
     }
 }
 
@@ -862,6 +862,7 @@ fn rustdoc(cx: &mut Context<'_, '_>, unit: &Unit) -> CargoResult<Work> {
                 },
                 false,
             )
+            .map_err(verbose_if_simple_exit_code)
             .with_context(|| format!("could not document `{}`", name));
 
         if let Err(e) = result {

--- a/tests/testsuite/doc.rs
+++ b/tests/testsuite/doc.rs
@@ -2640,3 +2640,24 @@ fn link_to_private_item() {
         )
         .run();
 }
+
+#[cargo_test]
+fn rustdoc_failure_hides_command_line_by_default() {
+    let p = project().file("src/lib.rs", "invalid rust code").build();
+
+    let string_to_test = "\
+Caused by:
+  process didn't exit successfully[..]rustdoc[..]";
+
+    // `cargo doc` doesn't print the full command line on failures by default
+    p.cargo("doc")
+        .with_stderr_does_not_contain(string_to_test)
+        .with_status(101)
+        .run();
+
+    // ... but it still does so if requested with `--verbose`.
+    p.cargo("doc --verbose")
+        .with_stderr_contains(string_to_test)
+        .with_status(101)
+        .run();
+}

